doc/adr006-modules.lit.md

```diff
@@ -0,0 +1,553 @@
+
+```ts
+export interface ValueDefinition {
+  /* Same as QuintDef kinds */
+  kind: ValueDefinitionKind
+  /* The name given to the defined operator */
+  identifier: string
+  /* Expression or definition id from where the name was collected */
+  reference?: bigint
+  /* Optional scope, an id pointing to the QuintIr node that introduces the name */
+  scope?: bigint
+  /* Optional type annotation */
+  typeAnnotation?: QuintType
+}
+```
+
+The most obvious issue is that names are not unique identifiers in a module,
+since operators may have parameters and nested definitions that carry the same
+names. This is illustrated by the definitions of `inc` and `dec` in the module
+`fun`.
+
+To deal with this issue, `ValueDefinition` contains an optional identifier
+called `scope`. This identifier is used to filter out the potential name
+clashes. This filtering happens on every lookup, which potentially slows down
+large specification. We could have done it once in the name resolution stage.
+This is tracked in the issue
+[#625](https://github.com/informalsystems/quint/issues/625).
+
+#### 2.3.3. D3: No distinction between variables produced by multiple instances
+
+The lookup tables are currently tuned towards particular kinds of analysis such
+as type checking. They are not general enough to work other kinds of stages
+such as simulation: For simulation and translation to Apalache, it is important
+that `I3::x` and `I5::x` point to two different versions of `x`, whereas they
+currently point to the same variable.
+
+#### 2.3.4. D4: Confusing syntax of module instances
+
+Recall how we produced new instances in our example:
+
+```bluespec
+module general {
+  module I3 = general(N = 3)
+  module I5 = general(N = 5)
+  ...
+}
+```
+
+In the earlier versions of the language, `I3` and `I5` were treated as modules.
+We have decided to remove support for nested modules in the issue
+[#548](https://github.com/informalsystems/quint/issues/548). As a consequence
+of this decision, the syntax of instances became surprising: It refers to the
+concept that is not supported by the language anymore.
+
+The most obvious approach would be to replace the keyword `module` with
+`instance`. However, the keyword `instance` seems to produce false expectations
+that instances behave like objects. In Quint (as well as in TLA+), an instance
+is a copy of a module, in which some names are substituited with expressions.
+We feel that the keywords such as `instance` and `extends` are heavily
+overloaded in the PL world, so it would be better to avoid them.
+
+## 3. Options
+
+### Make IR classes rich and use references
+
+A straightforward solution to the issue D2 would be to introduce the results of
+the analyses directly into the intermediate representation. For instance, we
+could extend `QuintName` with an optional type, an optional effect, an optional
+reference to the definition that the name refers to:
+
+```ts
+export interface QuintName extends WithId {
```

Comment:
In order for name resolution to "fill" the `_ref` fields, type checker "fill" the `_type` field, etc. We would have to make this representation mutable somehow, right? I feel like we might be going towards a direction of having a mutable IR anyways (i.e. to implement flattening), and that seems ok to me. If we can assign types and effects to all components of the IR, then we don't need the `Map<bigint, T>` structures anymore.

Reply:
I would really like to avoid such a rich data structure. It is typically much harder to serialize and debug. Also, it's not that obvious which stages modify which parts of the IR in this case.

Reply:
Actually, we might want different approaches for this "filling" and for the flattening. For filling, we could but all these new optional fields under a mutable object in a `analysisData` field, and then do things like `expr.analysisData.addType(inferredType)` to "update" the IR.

For flattening, we just have to add/remove definitions, so we could directly modify the `defs` array from the module.

Reply:
> I would really like to avoid such a rich data structure. It is typically much harder to serialize and debug. Also, it's not that obvious which stages modify which parts of the IR in this case.

I'm not sure I understand. Are you suggesting we implement this "rich" data structure you suggested on the ADR but don't make it mutable? So we'd have to update it with data from `Map<bigint, T>` at some point(s)?

Reply:
I mean, this is under the options. So I did not intend to implement this one. I should probably make it clear though.